File: javascript/lib/job_sequence.js
var JobList = require('../lib/job_list.js').JobList

var JobSequence = function JobSequence(input) {
	this.input = input;
	this.sequence = [];

	this.job_list = function() {
		return new JobList(input);
	}
	
	this.validate_input = function() {
		if(this.job_list().self_dependency_exists()) throw new Error("jobs cannot depend upon themselves");
		if(this.job_list().circular_dependency_exists()) throw new Error("input string cannot create a circular dependency");
	}
	
	this.output = function() {
		this.validate_input();
		this.calculate();
		
		return this.sequence.join();
	} 
	
	this.delete = function(job) {
		var dependent_position = this.sequence.indexOf(job);
		if(dependent_position > -1) this.sequence.splice(dependent_position,1);
	}
	
	this.add = function(name) {
	  this.sequence.push(name);
	}
	
	this.add_before = function(name, dependency) {
		var dependent_position = this.sequence.indexOf(name);
		this.sequence.splice(dependent_position, 0, dependency);				  
	}
	
	this.calculate = function() {
		var jobs = this.job_list().jobs();
		
		for(var i = 0; i < jobs.length; i++) {
			var job = jobs[i];
			
			this.add(job.name);

			if(job.dependency) {
		    this.delete(job.dependency);		    
		    this.add_before(job.name, job.dependency)
			}
		}
	}
}

exports.JobSequence = JobSequence;

File: ruby/lib/job_sequence.rb
require 'lib/job'

class JobSequence
  attr_accessor :input

  def self.calculate(input)
    js = self.new(input)
    js.result
  end 

  def initialize(input)
    self.input = input
  end

  def result
    raise ArgumentError, "Job cannot depend upon themselves" if self_dependency_exists?
    raise ArgumentError, "Job sequence contains a circular dependency" if circular_dependency_exists?

    sequence.join
  end

  private
  def dependents(target, list = [])
    jobs.each do |job|
      next unless job.dependency == target.name
      list.push(job.name)
      list += dependents(job, list) unless list.include?(job.dependency)
    end

    list
  end

  def depends_upon(target, list = [])
    jobs.each do |job|
      next unless job.name == target.dependency
      list.push(job.name)
      list += depends_upon(job, list) unless list.include?(job.dependency)
    end

    list
  end

  def jobs
    self.input.collect do |line|
      Job.new(line)
    end
  end

  def self_dependency_exists?
    jobs.any? do |job|
      job.dependency == job.name
    end
  end

  def circular_dependency_exists?
    jobs.any? do |job|
      (dependents(job) & depends_upon(job)).size > 0
    end
  end

  def sequence
    jobs.reduce([]) do |sequence, job|
      sequence.push(job.name) unless sequence.include?(job.name)

      if job.dependency
        sequence.delete(job.dependency)

        dependent_position = sequence.index(job.name)
        sequence.insert(dependent_position, job.dependency)
      end

      sequence
    end
  end
end
